Add Dag

eanylin opened the pull request:
Add basic OpenStack CLI call dag

eanylin committed:
Add Dag

Add basic OpenStack CLI call dag

alanmeadows commented on shipyard_airflow/dags/openstack_api_call.py:
```diff
@@ -0,0 +1,47 @@
+# airflow to perform OpenStack API calls
+t2 = BashOperator(
+    task_id='nova_list',
+    bash_command='source ' + os.getcwd() + '/dags/admin-openrc.sh' + ';' + 'nova' + ' list',
```
This is where an operator would shine, hiding complexity like sourcing openrc.

eanylin replied:
@alanmeadows Agreed.  I have re-written the codes using an OpenStack Operator instead.  The OpenStack Operator will be a plugin that can be called upon by Airflow and be consumed by the dag.

fmontei commented on shipyard_airflow/dags/openstack_api_call.py:
```diff
@@ -0,0 +1,47 @@
+# airflow to perform OpenStack API calls
+t2 = BashOperator(
+    task_id='nova_list',
+    bash_command='source ' + os.getcwd() + '/dags/admin-openrc.sh' + ';' + 'nova' + ' list',
```
Might be better to use:
`'source ' + os.path.join(os.getcwd(), 'dags/admin-operator.sh') + ...`

https://docs.python.org/2/library/os.path.html#os.path.join

eanylin replied:
@fmontei Thanks for the feedback.  I have kind of re-write what I posted initially.

fmontei commented on shipyard_airflow/dags/openstack_api_call.py:
```diff
@@ -0,0 +1,47 @@
+    'owner': 'airflow',
+    'depends_on_past': False,
+    'start_date': datetime(2017, 6, 14),
+    'email': ['airflow@airflow.com'],
```
Is the email 'airflow@airflow.com' a left-over dummy value or intentiona;?

fmontei commented on shipyard_airflow/dags/openstack_api_call.py:
```diff
@@ -0,0 +1,47 @@
+
+t3 = BashOperator(
+    task_id='neutron_net_list',
+    bash_command='source ' + os.getcwd() + '/dags/admin-openrc.sh' + ';' + 'neutron' + ' net-list',
```
See above.

eanylin committed:
Add OpenStack Operator

Define as a plugin so that Airflow is able to
pick up the plugin and all the operators it define

eanylin committed:
Update Dag for openrc path

alanmeadows commented on shipyard_airflow/dags/openstack_api_call.py:
```diff
@@ -0,0 +1,78 @@
+    bash_command='date',
+    dag=dag)
+
+## Note that the openrc.sh file needs to be placed on a volume that can be
```
Long-term, you need to retrieve these details like this from deckhand /cc @fmontei

eanylin replied:
Agree.  The information will need to come from deckhand in the long term.

alanmeadows commented on shipyard_airflow/plugins/openstack_operators.py:
```diff
@@ -0,0 +1,88 @@
+        logging.info("Output:")
+
+        line = ''
+        for line in iter(openstack_cli.stdout.readline, b''):
```
Does capturing and displaying the logging here interfere with airflows own built in STDOUT capturing mechanism for all external calls or is the logging module magically playing nice with how airflow wants to log?

I just want to make sure we're doing it the "proper" way so we only have to worry about log aggregation/collection in one place -- where airflow is sending everything.

eanylin replied:
Actually I had looked at the existing codes from the operators and followed their logic.  Hence I can see that the logs are shown in airflow UI under the log tab.  For instance, a sample output will look like the below:

```
[2017-06-23 03:11:24,350] {models.py:168} INFO - Filling up the DagBag from /usr/local/airflow/dags/openstack_api_call.py
[2017-06-23 03:11:24,945] {base_task_runner.py:112} INFO - Running: ['bash', '-c', u'airflow run openstack_cli service_list_task 2017-06-23T03:09:57.568664 --job_id 5 --raw -sd DAGS_FOLDER/openstack_api_call.py']
[2017-06-23 03:11:26,081] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,079] {__init__.py:57} INFO - Using executor CeleryExecutor
[2017-06-23 03:11:26,165] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,164] {driver.py:120} INFO - Generating grammar tables from /usr/lib/python2.7/lib2to3/Grammar.txt
[2017-06-23 03:11:26,208] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,204] {driver.py:120} INFO - Generating grammar tables from /usr/lib/python2.7/lib2to3/PatternGrammar.txt
[2017-06-23 03:11:26,468] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,467] {models.py:168} INFO - Filling up the DagBag from /usr/local/airflow/dags/openstack_api_call.py
[2017-06-23 03:11:26,987] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:26,987] {models.py:1128} INFO - Dependencies all met for <TaskInstance: openstack_cli.service_list_task 2017-06-23 03:09:57.568664 [queued]>
[2017-06-23 03:11:27,006] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,005] {models.py:1128} INFO - Dependencies all met for <TaskInstance: openstack_cli.service_list_task 2017-06-23 03:09:57.568664 [queued]>
[2017-06-23 03:11:27,006] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,005] {models.py:1328} INFO - 
[2017-06-23 03:11:27,006] {base_task_runner.py:95} INFO - Subtask: --------------------------------------------------------------------------------
[2017-06-23 03:11:27,007] {base_task_runner.py:95} INFO - Subtask: Starting attempt 1 of 2
[2017-06-23 03:11:27,007] {base_task_runner.py:95} INFO - Subtask: --------------------------------------------------------------------------------
[2017-06-23 03:11:27,007] {base_task_runner.py:95} INFO - Subtask: 
[2017-06-23 03:11:27,018] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,018] {models.py:1352} INFO - Executing <Task(OpenStackOperator): service_list_task> on 2017-06-23 03:09:57.568664
[2017-06-23 03:11:27,061] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,060] {openstack_operators.py:45} INFO - Running OpenStack Command: openstack service list
[2017-06-23 03:11:27,098] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:27,097] {openstack_operators.py:59} INFO - Output:
[2017-06-23 03:11:30,699] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - +----------------------------------+----------+----------+
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | ID                               | Name     | Type     |
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - +----------------------------------+----------+----------+
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | 3665ec03bb4045f6a33226923a0b283d | nova     | compute  |
[2017-06-23 03:11:30,700] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | 6ef93b0dda6c45c2b95cd0302e72fe2a | glance   | image    |
[2017-06-23 03:11:30,701] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,699] {openstack_operators.py:64} INFO - | 7cf15d3215014d148c51c9b57ae85bb0 | keystone | identity |
[2017-06-23 03:11:30,702] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,700] {openstack_operators.py:64} INFO - | a20df35688944634965682c55f6135ea | neutron  | network  |
[2017-06-23 03:11:30,702] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,700] {openstack_operators.py:64} INFO - +----------------------------------+----------+----------+
[2017-06-23 03:11:30,749] {base_task_runner.py:95} INFO - Subtask: [2017-06-23 03:11:30,747] {openstack_operators.py:70} INFO - Command exited with return code 0
[2017-06-23 03:11:34,979] {jobs.py:2107} INFO - Task exited with return code 0
```

alanmeadows commented on shipyard_airflow/dags/openstack_api_call.py:
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
```
We should probably start our own dags/samples directory as this isn't really a "real" dag?

eanylin replied:
Agree.  I have moved it under the samples directory.

eanylin committed:
Move sample dag to samples folder

eanylin committed:
Add Operator to Retrieve Task State

eanylin committed:
Update Error Handling Logic for XCOM Response

eanylin committed:
Add rest_api_plugin

Codes are derived from https://github.com/teamclairvoyant/airflow-rest-api-plugin

There is a need to augment the existing Airflow REST API capabilities
due to its limited functionalities.  As such, a wrapper of the Airflow
CLI is much needed.

mark-burnett commented on shipyard_airflow/plugins/openstack_operators.py:
```diff
@@ -0,0 +1,88 @@
+        pipe = subprocess.Popen(". %s; env" % self.openrc_file, stdout=subprocess.PIPE, shell=True)
+        data = pipe.communicate()[0]
+        env = dict((line.split("=", 1) for line in data.splitlines()))
+        os.environ.update(env)
```
Does this get cleared later?

Regardless, I guess you should consider passing environment on to the next Popen call below, which should avoid the possibility of leaving dangling environment (and possibly credentials) in this process.

eanylin replied:
I agree that we should pass the environment on to the next Popen call as it is better that way.  I have updated the codes to do that.

mark-burnett commented on shipyard_airflow/plugins/openstack_operators.py:
```diff
@@ -0,0 +1,88 @@
+
+
+        # Execute the OpenStack CLI Command
+        openstack_cli = subprocess.Popen(shlex.split(self.openstack_command), stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
As suggested above, you can use something like:

`subprocess.Popen(..., env=my_env_dict, ...)`

instead of modifying the environment of the currently running process.

eanylin replied:
Updated.  Thanks Mark for the suggestion.

mark-burnett commented on shipyard_airflow/plugins/openstack_operators.py:
```diff
@@ -0,0 +1,88 @@
+
+
+        # Execute the OpenStack CLI Command
+        openstack_cli = subprocess.Popen(shlex.split(self.openstack_command), stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
Is it necessary to take the `openstack_command` as a string instead of a list?  It seems like having to split it into a list with `shlex.split` is a bit awkward.

eanylin replied:
I have changed it into a list so that we do not need to use shlex.split

mark-burnett commented on shipyard_airflow/plugins/airflow_task_state_operators.py:
```diff
@@ -0,0 +1,84 @@
+        logging.info("Running Airflow Command: " + self.airflow_command)
+
+        # Execute Airflow CLI Command
+        airflow_cli = subprocess.Popen(shlex.split(self.airflow_command), stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```
Here, we probably don't need to use `shlex.split`, because we were given the separate pieces of the airflow command, so we can just use them directly to make the list.

eanylin replied:
Agree.  Have updated.

eanylin committed:
Update Airflow Dag and Plugin

Updated codes based on Feedbacks

mark-burnett commented on shipyard_airflow/plugins/openstack_operators.py:
```diff
@@ -42,18 +41,15 @@ def __init__(self,
 
     def execute(self, context):
-        logging.info("Running OpenStack Command: " + self.openstack_command)
+        logging.info("Running OpenStack Command: " + ' '.join(self.openstack_command))
```
I'm personally ok with just logging the list too:

```
logging.info('Running OpenStack Command: %s', self.openstack-command)
```

eanylin replied:
Ok. I have updated it to log list instead.

mark-burnett commented on shipyard_airflow/plugins/openstack_operators.py:
```diff
@@ -42,18 +41,15 @@ def __init__(self,
 
-        # Emulate "source" in bash. Sets up environment variables.
+        # Build environment variables.
         pipe = subprocess.Popen(". %s; env" % self.openrc_file, stdout=subprocess.PIPE, shell=True)
```
Do we need to execute this file, or can we just read it and assume/assert it's all environment variables?

eanylin replied:
The current assumption is that the rc file will contain only the environment information that is needed to access the OpenStack services.  Hence we will just need read that information.

mark-burnett commented on shipyard_airflow/plugins/rest_api_plugin.py:
```diff
@@ -435,7 +435,6 @@ def index(self):
     # '/api' REST Endpoint where API requests should all come in
     @csrf.exempt  # Exempt the CSRF token
     @expose('/api', methods=["GET", "POST"])
-    @http_token_secure  # On each request,
```
Is this change unrelated to review?

eanylin replied:
Yes, this is not related.  I found a bug when I was testing (had forgotten to remove this line initially and it was causing problems).

mark-burnett commented on shipyard_airflow/plugins/openstack_operators.py:
```diff
@@ -42,18 +41,15 @@ def __init__(self,
 
-        # Emulate "source" in bash. Sets up environment variables.
+        # Build environment variables.
         pipe = subprocess.Popen(". %s; env" % self.openrc_file, stdout=subprocess.PIPE, shell=True)
```
If I understand this correctly, we will fork out and "source" this rc file for every step of a DAG that connects to OpenStack?

We may want to change that later.

eanylin replied:
Thats the idea for now.  But this will not be the end state as the information will need to come from DeckHand rather than a rc file.  The integration is not ready at the moment.

mark-burnett commented on shipyard_airflow/plugins/rest_api_plugin.py:
```diff
@@ -0,0 +1,744 @@
+# Metadata about the APIs and how to call them. Representing them like this allows us to dynamically generate the APIs
+# in the admin page and dynamically execute them. This also allows us to easily add new ones.
+# API Object definition is described in the comment block above.
+apis_metadata = [
```
It might be nice to move this into a json file out of the way of the code, if the intent is for it to remain data.

eanylin replied:
Actually I am re-using the codes and have only removed certain parts of the codes that I felt was not relevant for our case.  I am only interested in the results of stdout and stderr in most cases (this piece of codes from GitHub was meant to wrap the Airflow CLI that has much more functionalities than the current experimental Airflow REST API).

mark-burnett reviewed (approved):
Thanks for addressing my initial comments.  I think the remaining feedback is less critical.

eanylin committed:
Update Logging to List
